File: skills/anti-ai-filter/check.py
#!/usr/bin/env python3
"""
Anti-AI Filter check script.

Mechanical evaluator that runs after the writer-agent returns a draft.
Replaces the vibes-based "final pass" — the agent can't claim to have run
this without actually invoking it.

Usage:
    python3 check.py <draft.md> --format <format-tier> [--voice-anchor <path>]

Format tiers:
    reddit-post-incident    5-15% caps target (lead with scene)
    reddit-post-essay       50-70% caps target (lead with claim)
    reddit-reply            5-15% caps target
    chat                    5-15% caps target
    blog-post               70-90% caps target
    longform                70-90% caps target
    marketing               85-95% caps target
    email                   70-90% caps target

Exit code 0 = pass, 1 = fail. Failures printed to stdout in structured form
the agent can parse and act on.
"""

import argparse
import re
import sys
from pathlib import Path

FORMAT_TIERS = {
    "reddit-post-incident": (5, 15),
    "reddit-post-essay": (50, 70),
    "reddit-reply": (5, 15),
    "chat": (5, 15),
    "casual-reply": (5, 15),
    "blog-post": (70, 90),
    "longform": (70, 90),
    "marketing": (85, 95),
    "landing": (85, 95),
    "email": (70, 90),
    "email-cold": (70, 90),
    "email-warm": (70, 90),
}

EM_DASH_BUDGET_PER_500_WORDS = 1


def read_file(path: Path) -> str:
    return path.read_text(encoding="utf-8")


def strip_frontmatter(text: str) -> str:
    """Strip leading YAML frontmatter if present."""
    if text.startswith("---\n"):
        end = text.find("\n---\n", 4)
        if end != -1:
            return text[end + 5 :]
    return text


def get_body_paragraphs(text: str) -> list[str]:
    """Split into non-empty paragraphs (separated by blank lines)."""
    text = strip_frontmatter(text)
    paragraphs = re.split(r"\n\s*\n", text)
    # Skip headers, code blocks, frontmatter remnants
    body = []
    for p in paragraphs:
        p = p.strip()
        if not p:
            continue
        if p.startswith("#") or p.startswith("```") or p.startswith("|"):
            continue
        if p.startswith("---") or p == "---":
            continue
        body.append(p)
    return body


def count_sentence_starts(body_text: str) -> tuple[int, int, list[str]]:
    """Return (total_starts, capital_starts, examples_of_lowercase_starts)."""
    sentences = re.split(r"(?<=[.!?])\s+", body_text)
    sentences = [s.strip() for s in sentences if s.strip()]
    total = 0
    caps = 0
    lowercase_examples = []
    for s in sentences:
        if not s:
            continue
        first = s[0]
        if not first.isalpha():
            continue
        total += 1
        if first.isupper():
            caps += 1
        else:
            if len(lowercase_examples) < 3:
                lowercase_examples.append(s[:60] + ("..." if len(s) > 60 else ""))
    return total, caps, lowercase_examples


def check_phrase_blacklist(draft_body: str, blacklist_path: Path) -> list[str]:
    """Return list of blacklisted phrases found in the draft (case-insensitive)."""
    if not blacklist_path.exists():
        return []
    matches = []
    body_lower = draft_body.lower()
    for line in blacklist_path.read_text(encoding="utf-8").splitlines():
        line = line.strip()
        if not line.startswith("- "):
            continue
        phrase = line[2:].strip()
        # Strip surrounding quotes and trailing parenthetical
        phrase = phrase.strip('"').strip("'")
        phrase = re.sub(r"\s*\([^)]*\)\s*$", "", phrase)
        if not phrase or len(phrase) < 4:
            continue
        # Skip pattern entries (contain X / Y / [X] placeholders)
        if any(token in phrase for token in [" X ", " Y ", "[X]", "[Y]", "(X)", "(any"]):
            continue
        if phrase.lower() in body_lower:
            matches.append(phrase)
    return matches


def count_em_dashes(draft_body: str) -> int:
    return draft_body.count("—") + draft_body.count("–")


def check_voice_anchor_leak(draft_body: str, anchor_path: Path, min_words: int = 5) -> list[str]:
    """Return list of 5+ word substrings copied verbatim from voice anchor."""
    if not anchor_path.exists():
        return []
    anchor_text = strip_frontmatter(anchor_path.read_text(encoding="utf-8"))

    # Tokenize both into normalized words
    def normalize(text: str) -> list[str]:
        return re.findall(r"\b[\w']+\b", text.lower())

    anchor_words = normalize(anchor_text)
    draft_words = normalize(draft_body)

    # Build set of N-grams from anchor
    anchor_ngrams = set()
    for i in range(len(anchor_words) - min_words + 1):
        anchor_ngrams.add(" ".join(anchor_words[i : i + min_words]))

    # Scan draft for matches
    leaks = []
    for i in range(len(draft_words) - min_words + 1):
        ngram = " ".join(draft_words[i : i + min_words])
        if ngram in anchor_ngrams:
            leaks.append(ngram)
    return list(dict.fromkeys(leaks))  # Dedupe, preserve order


def check_rhythm(paragraphs: list[str]) -> list[str]:
    """Return list of paragraphs lacking at least one rambling sentence."""
    failures = []
    for i, p in enumerate(paragraphs):
        sentences = re.split(r"(?<=[.!?])\s+", p)
        rambling = False
        for s in sentences:
            words = re.findall(r"\b[\w']+\b", s)
            commas_without_breaks = s.count(",") - s.count(":") - s.count(";")
            if len(words) >= 60 or commas_without_breaks >= 3:
                rambling = True
                break
        if not rambling and len(p) > 200:  # Only flag substantial paragraphs
            failures.append(f"P{i+1}: {p[:80]}{'...' if len(p) > 80 else ''}")
    return failures


def main():
    parser = argparse.ArgumentParser(description="Anti-AI filter check")
    parser.add_argument("draft", type=Path, help="Path to draft file")
    parser.add_argument(
        "--format",
        required=True,
        choices=list(FORMAT_TIERS.keys()),
        help="Format tier (determines casing target)",
    )
    parser.add_argument("--voice-anchor", type=Path, help="Optional voice anchor file to check for leakage")
    parser.add_argument(
        "--blacklist",
        type=Path,
        default=Path(__file__).parent / "phrase-blacklist.md",
        help="Path to phrase-blacklist.md",
    )
    parser.add_argument(
        "--blacklist-local",
        type=Path,
        help="Optional path to phrase-blacklist-local.md",
    )
    args = parser.parse_args()

    if not args.draft.exists():
        print(f"ERROR: draft file not found: {args.draft}", file=sys.stderr)
        sys.exit(2)

    draft_text = read_file(args.draft)
    body = strip_frontmatter(draft_text)
    paragraphs = get_body_paragraphs(draft_text)
    body_text = "\n\n".join(paragraphs)
    word_count = len(re.findall(r"\b[\w']+\b", body_text))

    failures = []
    summary_lines = []

    print("=== Anti-AI Filter Check ===")
    print(f"Draft: {args.draft}")
    print(f"Format tier: {args.format}")
    print(f"Word count: {word_count}")
    print()

    # 1. Phrase blacklist
    blacklist_hits = check_phrase_blacklist(body_text, args.blacklist)
    if args.blacklist_local:
        blacklist_hits.extend(check_phrase_blacklist(body_text, args.blacklist_local))
    if blacklist_hits:
        print(f"PHRASE BLACKLIST: FAIL ({len(blacklist_hits)} matches)")
        for hit in blacklist_hits:
            print(f"  - {hit}")
        failures.append(f"phrase-blacklist: {len(blacklist_hits)} match(es)")
    else:
        print("PHRASE BLACKLIST: PASS")
    print()

    # 2. Casing
    target_min, target_max = FORMAT_TIERS[args.format]
    total_starts, caps_starts, lowercase_examples = count_sentence_starts(body_text)
    if total_starts == 0:
        pct = 0
    else:
        pct = round(100 * caps_starts / total_starts)
    print(f"CASING: target {target_min}-{target_max}% caps, measured {pct}% ({caps_starts}/{total_starts} starts)")
    if total_starts > 0 and (pct < target_min - 5 or pct > target_max + 5):
        print("  FAIL: outside target range")
        if pct < target_min:
            print(f"  needs: {(target_min - pct) * total_starts // 100 + 1} more capital sentence-start(s)")
        else:
            print(f"  needs: {(pct - target_max) * total_starts // 100 + 1} fewer capital sentence-start(s)")
        failures.append(f"casing: {pct}% (target {target_min}-{target_max}%)")
    elif total_starts > 0 and (pct < target_min or pct > target_max):
        print(f"  WARN: just outside target (within ±5pp tolerance)")
    else:
        print("  PASS")
    print()

    # 3. Em-dashes
    em_count = count_em_dashes(body_text)
    em_budget = max(1, word_count // 500)
    print(f"EM-DASHES: count={em_count}, budget={em_budget} (per {EM_DASH_BUDGET_PER_500_WORDS}/500 words)")
    if em_count > em_budget:
        print(f"  FAIL: over budget by {em_count - em_budget}")
        failures.append(f"em-dashes: {em_count} (budget {em_budget})")
    else:
        print("  PASS")
    print()

    # 4. Voice anchor leak
    if args.voice_anchor:
        leaks = check_voice_anchor_leak(body_text, args.voice_anchor)
        print(f"VOICE ANCHOR LEAK: {len(leaks)} match(es) ≥5 words")
        if leaks:
            for leak in leaks:
                print(f'  - "{leak}"')
            failures.append(f"voice-anchor-leak: {len(leaks)} 5+word match(es)")
        else:
            print("  PASS")
    else:
        print("VOICE ANCHOR LEAK: skipped (no --voice-anchor provided)")
    print()

    # 5. Rhythm
    rhythm_failures = check_rhythm(paragraphs)
    if rhythm_failures:
        print(f"RHYTHM: FAIL ({len(rhythm_failures)} paragraph(s) without rambling sentence)")
        for f in rhythm_failures:
            print(f"  - {f}")
        failures.append(f"rhythm: {len(rhythm_failures)} paragraph(s)")
    else:
        print("RHYTHM: PASS (all body paragraphs have ≥1 rambling sentence or are short)")
    print()

    # Overall
    if failures:
        print("=" * 40)
        print(f"OVERALL: FAIL ({len(failures)} issue(s))")
        for f in failures:
            print(f"  - {f}")
        print()
        print("Rewrite the draft to address each failure, then re-run this check.")
        sys.exit(1)
    else:
        print("=" * 40)
        print("OVERALL: PASS")
        sys.exit(0)


if __name__ == "__main__":
    main()

File: skills/anti-ai-filter/check.mjs
#!/usr/bin/env node
/**
 * Anti-AI Filter check script.
 *
 * Mechanical evaluator that runs after the writer-agent returns a draft.
 * Replaces the vibes-based "final pass" - the agent can't claim to have run
 * this without actually invoking it.
 *
 * Zero dependencies. Uses only Node.js standard library. Anyone running
 * Claude Code already has Node, so nothing extra to install.
 *
 * Usage:
 *   node check.mjs <draft.md> --format <format-tier> [--voice-anchor <path>]
 *
 * Format tiers (caps target):
 *   reddit-post-incident   5-15%   (lead with scene)
 *   reddit-post-essay      50-70%  (lead with claim)
 *   reddit-reply           5-15%
 *   chat                   5-15%
 *   blog-post              70-90%
 *   longform               70-90%
 *   marketing              85-95%
 *   email                  70-90%
 *
 * Exit code 0 = pass, 1 = fail. Failures printed to stdout in structured
 * form the agent can parse and act on.
 */

import { readFileSync, existsSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));

const FORMAT_TIERS = {
  "reddit-post-incident": [5, 15],
  "reddit-post-essay": [50, 70],
  "reddit-reply": [5, 15],
  chat: [5, 15],
  "casual-reply": [5, 15],
  "blog-post": [70, 90],
  longform: [70, 90],
  marketing: [85, 95],
  landing: [85, 95],
  email: [70, 90],
  "email-cold": [70, 90],
  "email-warm": [70, 90],
};

// ---------------- arg parsing ----------------

function parseArgs(argv) {
  const args = { _: [], format: null, voiceAnchor: null, blacklist: null, blacklistLocal: null };
  for (let i = 2; i < argv.length; i++) {
    const a = argv[i];
    if (a === "--format") args.format = argv[++i];
    else if (a === "--voice-anchor") args.voiceAnchor = argv[++i];
    else if (a === "--blacklist") args.blacklist = argv[++i];
    else if (a === "--blacklist-local") args.blacklistLocal = argv[++i];
    else if (a === "-h" || a === "--help") {
      printHelp();
      process.exit(0);
    } else args._.push(a);
  }
  return args;
}

function printHelp() {
  console.log(`Anti-AI Filter check

Usage: node check.mjs <draft.md> --format <tier> [--voice-anchor <path>]

Format tiers: ${Object.keys(FORMAT_TIERS).join(", ")}

Optional:
  --voice-anchor <path>     File to check for 5+ word verbatim copies
  --blacklist <path>        Override default phrase-blacklist.md path
  --blacklist-local <path>  Additional local blacklist to check`);
}

// ---------------- helpers ----------------

function readFile(path) {
  return readFileSync(path, "utf8");
}

function stripFrontmatter(text) {
  if (text.startsWith("---\n")) {
    const end = text.indexOf("\n---\n", 4);
    if (end !== -1) return text.slice(end + 5);
  }
  return text;
}

function getBodyParagraphs(text) {
  const stripped = stripFrontmatter(text);
  const paragraphs = stripped.split(/\n\s*\n/);
  const body = [];
  for (let p of paragraphs) {
    p = p.trim();
    if (!p) continue;
    if (p.startsWith("#") || p.startsWith("```") || p.startsWith("|")) continue;
    if (p.startsWith("---")) continue;
    body.push(p);
  }
  return body;
}

function countWords(text) {
  const matches = text.match(/\b[\w']+\b/g);
  return matches ? matches.length : 0;
}

function countSentenceStarts(bodyText) {
  const sentences = bodyText.split(/(?<=[.!?])\s+/).map((s) => s.trim()).filter(Boolean);
  let total = 0;
  let caps = 0;
  const lowercaseExamples = [];
  for (const s of sentences) {
    if (!s) continue;
    const first = s[0];
    if (!/[a-zA-Z]/.test(first)) continue;
    total++;
    if (first === first.toUpperCase()) caps++;
    else if (lowercaseExamples.length < 3) {
      lowercaseExamples.push(s.length > 60 ? s.slice(0, 60) + "..." : s);
    }
  }
  return { total, caps, lowercaseExamples };
}

function checkPhraseBlacklist(bodyText, blacklistPath) {
  if (!existsSync(blacklistPath)) return [];
  const matches = [];
  const bodyLower = bodyText.toLowerCase();
  const blacklist = readFile(blacklistPath).split("\n");
  for (let line of blacklist) {
    line = line.trim();
    if (!line.startsWith("- ")) continue;
    let phrase = line.slice(2).trim();
    // Strip surrounding quotes and trailing parenthetical
    phrase = phrase.replace(/^["']|["']$/g, "");
    phrase = phrase.replace(/\s*\([^)]*\)\s*$/, "");
    if (!phrase || phrase.length < 4) continue;
    // Skip pattern entries (placeholders)
    if (/[ (]X[ )]|[ (]Y[ )]|\[X\]|\[Y\]/.test(phrase)) continue;
    if (bodyLower.includes(phrase.toLowerCase())) matches.push(phrase);
  }
  return matches;
}

function countEmDashes(bodyText) {
  return (bodyText.match(/—/g) || []).length + (bodyText.match(/–/g) || []).length;
}

function checkVoiceAnchorLeak(bodyText, anchorPath, minWords = 5) {
  if (!existsSync(anchorPath)) return [];
  const anchorText = stripFrontmatter(readFile(anchorPath));

  function normalize(text) {
    return (text.toLowerCase().match(/\b[\w']+\b/g) || []);
  }

  const anchorWords = normalize(anchorText);
  const draftWords = normalize(bodyText);

  const anchorNgrams = new Set();
  for (let i = 0; i <= anchorWords.length - minWords; i++) {
    anchorNgrams.add(anchorWords.slice(i, i + minWords).join(" "));
  }

  const seen = new Set();
  const leaks = [];
  for (let i = 0; i <= draftWords.length - minWords; i++) {
    const ngram = draftWords.slice(i, i + minWords).join(" ");
    if (anchorNgrams.has(ngram) && !seen.has(ngram)) {
      seen.add(ngram);
      leaks.push(ngram);
    }
  }
  return leaks;
}

function checkRhythm(paragraphs) {
  const failures = [];
  for (let i = 0; i < paragraphs.length; i++) {
    const p = paragraphs[i];
    if (p.length <= 200) continue; // only flag substantial paragraphs

    const sentences = p.split(/(?<=[.!?])\s+/);
    let rambling = false;
    for (const s of sentences) {
      const words = countWords(s);
      const commas = (s.match(/,/g) || []).length;
      const breaks = (s.match(/[:;]/g) || []).length;
      const commasWithoutBreaks = commas - breaks;
      if (words >= 60 || commasWithoutBreaks >= 3) {
        rambling = true;
        break;
      }
    }
    if (!rambling) {
      const preview = p.length > 80 ? p.slice(0, 80) + "..." : p;
      failures.push(`P${i + 1}: ${preview}`);
    }
  }
  return failures;
}

// ---------------- main ----------------

function main() {
  const args = parseArgs(process.argv);

  if (args._.length === 0) {
    console.error("ERROR: missing draft file. Use --help for usage.");
    process.exit(2);
  }
  if (!args.format) {
    console.error("ERROR: --format is required. Use --help for usage.");
    process.exit(2);
  }
  if (!FORMAT_TIERS[args.format]) {
    console.error(
      `ERROR: unknown format "${args.format}". Choices: ${Object.keys(FORMAT_TIERS).join(", ")}`,
    );
    process.exit(2);
  }

  const draftPath = args._[0];
  if (!existsSync(draftPath)) {
    console.error(`ERROR: draft file not found: ${draftPath}`);
    process.exit(2);
  }

  const blacklistPath = args.blacklist || join(__dirname, "phrase-blacklist.md");
  const draftText = readFile(draftPath);
  const paragraphs = getBodyParagraphs(draftText);
  const bodyText = paragraphs.join("\n\n");
  const wordCount = countWords(bodyText);

  const failures = [];

  console.log("=== Anti-AI Filter Check ===");
  console.log(`Draft: ${draftPath}`);
  console.log(`Format tier: ${args.format}`);
  console.log(`Word count: ${wordCount}`);
  console.log();

  // 1. Phrase blacklist
  let blacklistHits = checkPhraseBlacklist(bodyText, blacklistPath);
  if (args.blacklistLocal) {
    blacklistHits = blacklistHits.concat(checkPhraseBlacklist(bodyText, args.blacklistLocal));
  }
  if (blacklistHits.length > 0) {
    console.log(`PHRASE BLACKLIST: FAIL (${blacklistHits.length} matches)`);
    for (const hit of blacklistHits) console.log(`  - ${hit}`);
    failures.push(`phrase-blacklist: ${blacklistHits.length} match(es)`);
  } else {
    console.log("PHRASE BLACKLIST: PASS");
  }
  console.log();

  // 2. Casing
  const [targetMin, targetMax] = FORMAT_TIERS[args.format];
  const { total, caps } = countSentenceStarts(bodyText);
  const pct = total === 0 ? 0 : Math.round((100 * caps) / total);
  console.log(
    `CASING: target ${targetMin}-${targetMax}% caps, measured ${pct}% (${caps}/${total} starts)`,
  );
  if (total > 0 && (pct < targetMin - 5 || pct > targetMax + 5)) {
    console.log("  FAIL: outside target range");
    if (pct < targetMin) {
      const need = Math.floor(((targetMin - pct) * total) / 100) + 1;
      console.log(`  needs: ${need} more capital sentence-start(s)`);
    } else {
      const need = Math.floor(((pct - targetMax) * total) / 100) + 1;
      console.log(`  needs: ${need} fewer capital sentence-start(s)`);
    }
    failures.push(`casing: ${pct}% (target ${targetMin}-${targetMax}%)`);
  } else if (total > 0 && (pct < targetMin || pct > targetMax)) {
    console.log("  WARN: just outside target (within ±5pp tolerance)");
  } else {
    console.log("  PASS");
  }
  console.log();

  // 3. Em-dashes
  const emCount = countEmDashes(bodyText);
  const emBudget = Math.max(1, Math.floor(wordCount / 500));
  console.log(`EM-DASHES: count=${emCount}, budget=${emBudget} (1 per 500 words)`);
  if (emCount > emBudget) {
    console.log(`  FAIL: over budget by ${emCount - emBudget}`);
    failures.push(`em-dashes: ${emCount} (budget ${emBudget})`);
  } else {
    console.log("  PASS");
  }
  console.log();

  // 4. Voice anchor leak
  if (args.voiceAnchor) {
    const leaks = checkVoiceAnchorLeak(bodyText, args.voiceAnchor);
    console.log(`VOICE ANCHOR LEAK: ${leaks.length} match(es) ≥5 words`);
    if (leaks.length > 0) {
      for (const leak of leaks) console.log(`  - "${leak}"`);
      failures.push(`voice-anchor-leak: ${leaks.length} 5+word match(es)`);
    } else {
      console.log("  PASS");
    }
  } else {
    console.log("VOICE ANCHOR LEAK: skipped (no --voice-anchor provided)");
  }
  console.log();

  // 5. Rhythm
  const rhythmFailures = checkRhythm(paragraphs);
  if (rhythmFailures.length > 0) {
    console.log(`RHYTHM: FAIL (${rhythmFailures.length} paragraph(s) without rambling sentence)`);
    for (const f of rhythmFailures) console.log(`  - ${f}`);
    failures.push(`rhythm: ${rhythmFailures.length} paragraph(s)`);
  } else {
    console.log("RHYTHM: PASS (all body paragraphs have ≥1 rambling sentence or are short)");
  }
  console.log();

  // Overall
  if (failures.length > 0) {
    console.log("=".repeat(40));
    console.log(`OVERALL: FAIL (${failures.length} issue(s))`);
    for (const f of failures) console.log(`  - ${f}`);
    console.log();
    console.log("Rewrite the draft to address each failure, then re-run this check.");
    process.exit(1);
  } else {
    console.log("=".repeat(40));
    console.log("OVERALL: PASS");
    process.exit(0);
  }
}

main();
